distributed/stealing.py

```diff
@@ -464,6 +470,32 @@ def maybe_move_task(level, ts, sat, idl, duration, cost_multiplier):
             if s.digests:
                 s.digests["steal-duration"].add(stop - start)
 
+    def _maybe_pick_thief(self, ts, thieves):
+        """Try to be smart about picking a thief given some options. We're
+        trying to pick a thief which has dependencies for a given task, if
+        possible and will pick the one which works best for us given the
+        Scheduler.worker_objective
+
+        If no idle worker with dependencies is found, this returns None.
+        """
+        if ts._dependencies:
+            who_has = set()
+            for dep in ts._dependencies:
+                who_has.update(dep.who_has)
+
+            thieves_with_data = list(who_has & set(thieves))
+
+            # If there are potential thieves with dependencies we
+            # should prefer them and pick the one which works best.
+            # Otherwise just random/round robin
+            if thieves_with_data:
```

Comment:
If, say, just 1 idle worker holds dependencies, is it possible for that worker to get slammed with new tasks because of this? What has to happen between picking a worker as a thief and it getting removed from the idle set?

Reply:
> If, say, just 1 idle worker holds dependencies, is it possible for that worker to get slammed with new tasks because of this

Yes, this may hurt, particularly for sub-topologies like

```mermaid
flowchart BT
A --> B1
A --> B2
A --> B3
```

where A is very small and B* is very compute heavy, i.e. steal_ratio will always be small and tasks are even allowed to be stolen from "public", i.e. non-saturated workers. I believe for other cases, i.e. if tasks are just stolen from saturated workers, this should be OK since decide_worker is much smarter with initial task placement.

> What has to happen between picking a worker as a thief and it getting removed from the idle set?

Steal request must be confirmed which updates occupancies and recalculates the idle set. Somewhere on this PR a comment of mine is suggesting that this `_mabye_pick_thief` should incorporate `in_flight_occupancy` to reflect for the time until this happens.